validator/db/manage.go

```diff
@@ -89,53 +72,255 @@ func getPubKeyProposals(pubKey []byte, proposalsBucket *bolt.Bucket) (*pubKeyPro
 		return nil
 	}); err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "could not retrieve proposals for public key %x", pubKey[:12])
 	}
 
 	return &pubKeyProposals, nil
 }
 
-func createTargetStore(
+func createMergeTargetStore(
 	targetDirectory string,
 	allProposals []pubKeyProposals,
 	allAttestations []pubKeyAttestations) (err error) {
 
 	newStore, err := NewKVStore(targetDirectory, [][48]byte{})
 	defer func() {
+		errorMessage := "could not close the merged database"
 		if deferErr := newStore.Close(); deferErr != nil {
-			err = errors.Wrap(err, "Could not close the merged database")
+			if err != nil {
+				err = errors.Wrap(err, errorMessage)
+			} else {
+				err = errors.Wrap(deferErr, errorMessage)
+			}
+
 		}
 	}()
 	if err != nil {
-		return errors.Wrapf(err, "Could not initialize a new database in %s", targetDirectory)
+		return errors.Wrapf(err, "could not initialize a new database in %s", targetDirectory)
 	}
 
 	err = newStore.update(func(tx *bolt.Tx) error {
-		proposalsBucket := tx.Bucket(historicProposalsBucket)
+		allProposalsBucket := tx.Bucket(historicProposalsBucket)
 		for _, pubKeyProposals := range allProposals {
-			pubKeyBucket, err := proposalsBucket.CreateBucket(pubKeyProposals.PubKey)
+			proposalsBucket, err := createProposalsBucket(allProposalsBucket, pubKeyProposals.PubKey)
 			if err != nil {
-				return errors.Wrapf(err,
-					"Could not create proposals bucket for public key %x",
-					pubKeyProposals.PubKey[:12])
+				return err
 			}
-			for _, epochProposals := range pubKeyProposals.Proposals {
-				if err := pubKeyBucket.Put(epochProposals.Epoch, epochProposals.Proposals); err != nil {
-					return errors.Wrapf(err, "Could not add epoch proposals for epoch %v", epochProposals.Epoch)
-				}
+			if err := addEpochProposals(proposalsBucket, pubKeyProposals.Proposals); err != nil {
+				return err
 			}
 		}
 		attestationsBucket := tx.Bucket(historicAttestationsBucket)
 		for _, attestations := range allAttestations {
-			if err := attestationsBucket.Put(attestations.PubKey, attestations.Attestations); err != nil {
-				return errors.Wrapf(
-					err,
-					"Could not add public key attestations for public key %x",
-					attestations.PubKey[:12])
+			if err := addAttestations(attestationsBucket, attestations); err != nil {
+				return err
 			}
 		}
 		return nil
 	})
 
 	return err
 }
+
+func createSplitTargetStores(
+	targetDirectory string,
+	allProposals []pubKeyProposals,
+	allAttestations []pubKeyAttestations) (err error) {
+
+	var storesToClose []*Store
+	defer (func() {
+		errorMessage := "failed to close one or more stores"
+		failedToClose := false
+		for _, store := range storesToClose {
+			if deferErr := store.Close(); deferErr != nil {
+				failedToClose = true
+			}
+		}
+		if failedToClose {
+			if err != nil {
+				err = errors.Wrapf(err, errorMessage)
+			} else {
+				err = errors.New(errorMessage)
```

Comment:
You need to do something about these errors in this deferred function. The function will not run until createSplitTargetStores returns, so this error will have nowhere to do. I would suggest creating an error log here instead. There's no way to control nor extract the outputs of deferred functions

Reply:
Actually, there is:
https://github.com/prysmaticlabs/prysm/pull/6027#discussion_r432061440

Reply:
Wow awesome! Thanks for sharing, good thing to know :)